README.md
[phpSec](https://phpseclib.com/) - PHP security library
=======================================================
* phpSec is a open-source [PHP](http://php.net) security library that takes care
  of the common security tasks a web developer faces.

[![Build Status](https://travis-ci.org/phpsec/phpSec.png)](https://travis-ci.org/phpsec/phpSec)

[Official Website](https://phpseclib.com/) and [Documentation](http://phpseclib.com/docs)

Features
--------
* Data encryption
* XSS filter
* Password hashing
* Secure session handler
* CSRF protection
* Yubikey integration
* Authy integration
* Random data generator

Installing
---------------
phpSec is now a PSR-0 compatible library. this means that it can easilly be installed and loaded using [Composer](http://getcomposer.org/doc/00-intro.md).
You can also install phpSec manually, or using Git.

### Installing using [Composer](http://getcomposer.org/doc/00-intro.md)
To install using Composer just add phpSec to your composer.json file in your project directory.
```
{
    "require": {
        "phpsec/phpsec":"0.6.*"
    }
}
```

Then all you need to do is to run `$ php composer.phar install` .
phpSec can then be loaded using the Composer autoloader.

`require 'vendor/autoload.php';`

### Installing manually/Git
Download, checkout or peferrably add phpSec as a Git submodule. Then load the *lib/phpSec/bootstrap.php* file.
This is a really simple autoloader that will load phpSec classes when needed.

`require_once 'lib/phpSec/bootstrap.php';`

If you already have a PSR-0 compatible autoloader for your project there is no need to call the *bootstrap.php* file.
All you have to do is to register the *phpSec* namespace to the *phpSec/lib* folder.

If you want to add an autoloader to your project there is [one example here](http://gist.github.com/221634).
This can be initialized like this:

```php
<?php
require_once 'SplClassLoader.php';
$classLoader = new SplClassLoader('phpSec', '/var/www/vendor/phpSec/lib');
$classLoader->register();
```

For documentation on how to use the various phpSec functionality, take alook at the [phpsec/doc](https://github.com/phpsec/doc) repository. 

System requirements
-------------------
* PHP >= 5.3.7
* [Mcrypt](http://no.php.net/manual/en/mcrypt.installation.php), if you want to encrypt stuff.

Getting help / Contact
----------------------
 * [phpSec manual](https://github.com/phpsec/doc/)
 * [phpSec issues] (https://github.com/phpsec/phpSec/issues/)
 * [Twitter (@xqus)](http://twitter.com/xqus/)
 * [Website](https://phpseclib.com/)
 * E-mail: larsen@xqus.com

License
-------
phpSec is open-sourced software licensed under the MIT License.


CHANGELOG.md
phpSec changelog
================

0.6.1: February 25. 2013 - xqus
-------------------------------
* Clean up the way we start sessions.
* Add a timing safe comparison function (\phpSec\String\Compare).
* Use timing safe comparison when comparing hashes.
* Default password hashing method is now Bcrypt.

0.6.0: February 17. 2013 - xqus
-------------------------------
* [#111] Total rewrite: Now uses Pimple as DI container.
* [#113] Swap bcrypt prefix. Other small fixes to \phpSec\Crypt\Hash.
* Adds \phpSec\Http\Xfo, Implements HTTP X-Frame-Options.
* Adds \phpSec\Http\Hsts, Implements HTTP Strict Transport Security.
* Adds \phpSec\Http\Url, Impements URL manipulation prevention.
* Adds \phpSec\Auth\Mnemonic.php, Create easy to remember passphrases.
* [#107] phpSec throws exceptions instead of triggering errors.
* Updated tests.
* Various minor fixes.

0.5.1: December 12. 2012 - xqus
-------------------------------
* Fixes \phpSec\Text\Filter.

0.5.0: November 5. 2012 - xqus
------------------------------
* Added \phpSec\Auth\Google, Authenticate using Google Authenticator.
* Added \phpSec\Common\SplClassLoader, an improved autoloader.
* [#101] Added \phpSec\Common\Exec, method for executing programs in a PDO like way.
* Improved \phpSec\Common\Core\genUid().
* Added \phpSec\Crypt\Rand::bool(): Generate random boolean.
* Improved \phpSec\Crypt\Crypto. (Error messages and automatic padding).
* Fixed bug in session handler when setting custom encryption settings.
* [#100] Added support for Drupal and plain hashes in \phpSec\Crypt\Hash.
* Added \phpSec\Crypt\Hash::getInfo(), returns settings used to generate a hash.
* New class: \phpSec\String\Base32
* [#105] Core::getUid() should be static.

0.4.0: August 19. 2012 - xqus
-----------------------------
* Total restructure of library to comply with PSR-0.
* Version numbers now follows [Semantic Versioning](http://semver.org/)
* [#83] Added class \phpSec\Auth\Otpcard and moved PS OTP functionality from \phpSec\Auth\Otp to it.
* [#84, #85, #93, #94] Improvements to \phpSec\Auth\Otp.
* Fixed bug setting up storage on Windows platforms.
* New configuration option: Disable regeneration of session id for each request.
* [#91] Throw PHP notice if insecure pseudo-random data is generated.
* [#23] Added filename filter to \phpSec\Text\Filter::f().
* [#90] Improvements to pseudo-random data generator. (@ph-il)
* Adds support for [Authy](http://authy.com) authentication. 
* New autoloader.
* Code cleanup and small bug fixes.

beta-0.3: January 4. 2012 - xqus
--------------------------------
* [#66] phpSec can now run without storage configured.
* [#58] phpsecCrypt now pads data by default.
* [#81] phpsecPw now supports PBKDF2, and uses it as default.
* Added new password hashing class: phpsecHash. phpsecPw should not be used
  anymore.
* Fixed phpsecFilter::f().

beta-0.2: December 21. 2011 - xqus
----------------------------------
* [#43] Implemented PKCS7 padding in phpsecCrypt.
* [#62] Added phpsec::arrayCheck().
* [#63] Improved checking of json data passed to phpSec.
* Fixed call to error method in phpsecPw::check()
* [#51] phpsecPw now uses a binary salts.
* phpsecPw::age() removed.
* [#50] Session handler now uses a binary key to encrypt session data.
* [#42] Added mySQL support.
* Yubikey integration not considered experimental anymore.
  * [#48, #49] Improvements of the Yubikey integration.
  * [#75] Fixed: phpsecYubikey::getYubikeyId() could retrive wrong ID.
  * [#76] Fixed: phpsecYubikey::validOtp() to strict.
  * [#74] Fixed: phpsecYubikey::getResponse(): Number of attempts and timeout
          should be possible to configure.


beta-0.1: November 7. 2011 - xqus
---------------------------------
* [#33, #34]phpsec::f() is now phpsec::t(), and phpsec::f() is a simple XSS
  filter method.
* Removed the examples.php from source. See the online manual for examples:
  http://phpsec.xqus.com/manual
* Don't use session_regenerate_id() anymore, since it causes many bugs.
* Generate a stronger PHP session ID.
* [#30] Improve session hijacking protection.
* Improved the password hashing methods.
* Changed default encryption algorithm to RIJNDAEL-256.
* Added implemention of PBKDF2 as described in RFC 2898.
* Create a PBKDF2 MAC to ensure message integrity in phpsecCrypt. This breaks
  compability with older versions of phpSec.
* [#37] phpsecCrypt don't modify the encryption key anymore.
* Regenerate session ID in custom session handler.
* Stronger keys for encryption of session.
* Better error handling in phpSecCrypt.
* [#38] Storage class for general data storage for all sub-classes.
* [#41] Separate password, token and filter methods from core.
* Many minor fixes..

alpha-0.0.5: August 14. 2011 - xqus
-----------------------------------
* Added phpsecRand::arrayRand(): Method to select random keys from an array.
* Fixed bug in key generation method in phpsecCrypt().
  This means that data encrypted with phpSec alpha-0.0.4 and older will not
  decrypt in this version.

alpha-0.0.4: July 11. 2011 - xqus
---------------------------------
* [#25] Empty CSRF token creates hickups.
* Added initial GPG/PGP support. Still experimental.
* Added phpsecLog class to add better logging support.
  This means that phpsec::init() has to be called before setting the log dir:
  phpsecLog::$_logdir = 'filesystem:/var/www/phpSec/logs';
  Also phpsec::log() is now phpsecLog::log().
* Added syslog support to phpsecLog.
* phpsec::f() now accepts strings a array of data to filter.
* Improved error handling.
* phpsecCrypt::encrypt() performance improvements.
* Greatly improved key generation security in phpsecCrypt().
  This means that data encrypted with phpSec alpha-0.0.3 and older will not
  decrypt in this version.

alpha-0.0.3: March 9. 2011 - xqus
---------------------------------
* [#21] Use of phpSec session handler is now optional.

Common/Core.php
<?php namespace phpSec\Common;
/**
  phpSec - A PHP security library

  @author    Audun Larsen <larsen@xqus.com>
  @copyright Copyright (c) Audun Larsen, 2011, 2012
  @link      https://github.com/phpsec/phpSec
  @license   http://opensource.org/licenses/mit-license.php The MIT License
  @package   phpSec
 */

use \phpSec\Store;
use \phpSec\Crypt\Rand;

/**
 * phpSec core functionality.
 */
class Core {
  /**
   * Constant: Version number. Isn't really
   * used for something... yet.
   */
  const VERSION = '0.4.0';

  /**
   * Storage object.
   */
  public static $store = null;

  /**
   * User identifier.
   */
  public static $uid = null;

  /**
   * Provides the library a simple way of reporting errors to the developer using PHPs error
   * handler.
   *
   * @see http://php.net/manual/en/errorfunc.constants.php
   *
   * @param string $msg
   *   String containing the error message
   *
   * @param constant $level
   *   Error level (optional).
   *   If none is specified E_USER_WARNING is used.
   */
  public static function error($msg, $level = E_USER_WARNING) {
    $trace = debug_backtrace();
    trigger_error($msg.'. (Called from <strong>'.$trace[0]['file'].' line '.$trace[0]['line'].'</strong>)', $level);
  }

  /**
   * Check structure of an array.
   * This method checks the structure of an array (only the first layer of it) against
   * a defined set of rules.
   *
   * @param array $array
   *   Array to check.
   *
   * @param array $structure
   *   Expected array structure. Defined for example like this:
   *   array(
   *     'string' => array(
   *       'callback' => 'strlen',
   *       'params'   => array('%val'),
   *       'match'    => 3,
   *     ),
   *     'not allowed' = false, // Only makes sense with $strict = false
   *     'needed'      = true,
   *   ),
   *
   * @param bool $strict
   *   If strict is set to false we will allow keys that's not defined in the structure.
   */
  public static function arrayCheck($array, $structure, $strict = true) {
    $success = true;
    /* First compare the size of the two arrays. Return error if strict is enabled. */
    if(sizeof($array) != sizeof($structure) && $strict === true) {
      self::error('Array does not match defined structure');
      return false;
    }

    /* Loop trough all the defined keys defined in the structure. */
    foreach($structure as $key => $callbackArray) {
      if(isset($array[$key])) {
        /* The key exists in the array we are checking. */

        if(is_array($callbackArray) && isset($callbackArray['callback'])) {
          /* We have a callback. */

          /* Replace %val with the acutal value of the key. */
          $callbackArray['params'] = str_replace('%val', $array[$key], $callbackArray['params']);

          if(call_user_func_array($callbackArray['callback'], $callbackArray['params']) !== $callbackArray['match']) {
            /* Call the *duh* callback. If this returns false throw error, or an axe. */
            self::error('Array does not match defined structure. The '.$key.' key did not pass the '.$callbackArray['callback'].' callback');
            $success = false;
          }
        } elseif($callbackArray === false) {
          /* We don't have a callback, but we have found a disallowed key. */
          self::error('Array does not match defined structure. '.$key.' is not allowed');
          $success = false;
        }
      } else {
        /* The key don't exist in the array we are checking. */

        if($callbackArray !== false) {
          /* As long as this is not a disallowed key, sound the general alarm. */
          self::error('Array does not match defined structure. '.$key.' not defined');
          $success = false;
        }
      }
    }
    return $success;
  }

  /**
   * Returns a unique identifier in the format spsecified in
   * OpenID Authentication 2.0 protocol.
   * For example: 2005-05-15T17:11:51ZUNIQUE
   * This function is used to generate all unique tokens used by
   * phpSec.
   * @see http://openid.net/specs/openid-authentication-2_0.html
   *
   * @param integer $length
   *   The total length of the uid. Must be above 25.
   */
  public static function genUid($length = 50) {
    if($length < 25) {
      self::error('Length must be longer than 25');
      return false;
    }
    $timeStamp = gmdate('Y-m-d\TH:i:se');
    $randLength = $length-strlen($timeStamp);
    return $timeStamp.Rand::str($randLength);
  }

  public function getUid() {
  	/* Create a random token for each visitor and store it the users session.
       This is for example used to identify owners of cache data. */
    if(!isset($_SESSION['phpSec-uid'])) {
      $_SESSION['phpSec-uid'] = self::genUid();
    }

    return $_SESSION['phpSec-uid'];
  }

  public static function setStore($dsn) {

  	/* Open store. */
    list($storeType, $storeDest) = explode(':', $dsn, 2);
    switch($storeType) {
      case 'filesystem':
        self::$store = new Store\File($storeDest);
      break;
      case 'mysql':
        self::$store = new Store\Pdo($storeDest);
      break;
      default:
      self::error('Store type('.$storeType.') invalid', E_USER_ERROR);
    }
  }
}